docs/source/coding_guidelines.md

```diff
@@ -0,0 +1,213 @@
+    """
+```
+
+* End strings in error messages with a period.
+
+### Documentation
+
+We are using Sphinx for generating documentation. We recommend using a VSCode extension for auto-generating documentation templates, for example `njpwerner.autodocstring` (this extension is already pre-configured in VSCode's workspace settings file).
+
+Common mistakes when writing docstrings:
+
+* There must be a separating line between a function description and the documentation for its parameters.
+* In multi-line parameter descriptions, continuations on the next line must be indented.
+* Sphinx will merge the class description and the arguments of the constructor `__init__`. Hence, there is no need to
+  write any text in the constructor docstring other than the constructor arguments.
+* Use `>>>` to include code snippets. PyCharm will run intellisense on those to make authoring easier.
+* To generate the Sphinx documentation on your dev machine, run `make html` in the `./docs` folder, and then
+  open `./docs/build/html/index.html`
+
+Example:
+
+```python
+class Foo:
+    """
+    This is the class description.
+
+    The following block will be pretty-printed by Sphinx. Note the space between >>> and the code!
+    
+    Usage example:
+        >>> from module import Foo
+        >>> foo = Foo(bar=1.23)
+    """
+
+    ANY_ATTRIBUTE = "what_ever."
+    """Document class attributes after the attribute."""
+
+    def __init__(self, bar: float = 0.5) -> None:
+        """
+        :param bar: This is a description for the constructor argument.
+            Long descriptions should be indented.
+        """
+        self.bar = bar
+
+    def method(self, arg: int) -> None:
+        """
+        Method description, followed by an empty line. Argument names like ``arg`` are rendered nicely
+        if enclosed in double backtick.
+
+        This method can raise a :exc:`ValueError`.
+        
+        :param arg: This is a description for the method argument.
+            Long descriptions should be indented.
+        """
+```
+
+## Testing
+
+### Why do we write unit tests?
+
+If you think that there are sensitive parts in your code, where hidden errors can impact the project outcome, please take your time to protect parts of your code. As the team grows, so does the number of  users of the repository, and we may not be able to monitor changes on a frequent basis, but your tests can ensure that the expected behavior is preserved. We are not making it strictly mandatory, however users shall take the liability of untested code and its potential downstream impact.
+
+We do not need to test every part of the code depending on the project constraints, e.g., quality and test-coverage vs learning and hypothesis-testing. However, we expect a justification in each PR for this decision and it should be reviewed by team members. Please regularly monitor the code coverage results that are posted as comments on each pull request.
```

Comment:
It might not be very fair to ask to monitor the verbose coverage comments.

Reply:
I think this will go away with codecov

Reply:
It will :)